enhancements/network/vrf-cni.md

```diff
@@ -0,0 +1,123 @@
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+To allow isolation, and make it possible to have overlaps between the CIDRs of secondary networks (and between them and the pod's address space), we want to introduce a VRF meta cni plugin. Taking advantage of the chaining mechanism, it can be used to assign a secondary interface to a custom [VRF](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_and_managing_networking/configuring-virtual-routing-and-forwarding-vrf_configuring-and-managing-networking).
+
+## Motivation
+
+Telco customers maintain strict separation of multiple networks for different services or functionalities. When a device connects to multiple networks, the practice is to maintain strict separation of these networks.
+
+Moreover, the multiple networks a pod is connected to may have overlapping CIDRs, and those CIDRs can also overlap with the sdn’s CIDR. To maintain this kind of isolation, they typically use VRFs. Currently this is done by granting the containers the NET_ADMIN capabilities.
+
+### Goals
+
+Being able to assign a secondary interface to a particular vrf, and eventually having more secondary interfaces of the same pod assigned to the same VRF.
+
+### Non-Goals
+
+Assigning host level interfaces to VRFs.
+
+## Proposal
+
+The proposal is to introduce an additional CNI plugin that takes the vrf name as an input, and can be chained to the “main” cni used for the secondary interface, and that has the effect of creating the requested vrf (if it does not exist) and to add the interface to that vrf.
+
+In this way, the network interface returned by any CNI can be isolated in its own VRF, satisfying the above requirements without the need of additional capabilities.
+
+### User Stories [optional]
+
+#### Story 1
+
+As a user, I want to be able to optionally specify a VRF which the secondary nic is attached to.
+
+### Implementation Details/Notes/Constraints
+
+By adding several networks with the same `vrf` name, the CNI will ensure that they are all added to the same vrf, creating it only the first time.
+
+A `NetworkAttachmentDefinition` CNI config of `NetworkAttachmentDefinition` would look like
+
+```json
+{
+    "cniVersion": "0.3.1",
+    "name": "macvlan-vrf",
+    "plugins": [
+      {
+        "type": "macvlan",
+        "mode": "bridge",
+        "ipam": {
+          "type": "host-local",
+          "subnet": "192.168.1.0/24",
+          "rangeStart": "192.168.1.200",
+          "rangeEnd": "192.168.1.216",
+        }
+      },
+      {
+        "type": "vrf",
+        "vrfname": "blue"
+      }
+    ]
+}
+```
+
+This can be supported at `NetworkAttachmentDefinition` level, or by documenting (and validating) chained CNIs in the `Network` CRD of `operator.openshift.io/v1`.
+
+At the same time, the `SriovNetwork` type needs to be extended for chaining meta CNIs.
```

Comment:
Would it make sense to mention where the SriovNetwork API is defined, like: the `SriovNetwork` type in `SR-IOV Operator`

Reply:
I'll do